corehq/apps/reports/standard/cases/case_data.py

```diff
@@ -555,6 +561,255 @@ def close_case_view(request, domain, case_id):
     return HttpResponseRedirect(reverse('case_data', args=[domain, case_id]))
 
 
+MAX_CASE_COUNT = CASEBLOCK_CHUNKSIZE
+MAX_SUBCASE_DEPTH = 3
+
+
+def get_case_and_display_data(case_obj, domain):
+    """
+    Given a case object, recursively checks the case's related submission forms and for each form, the related
+    cases it has affected, and so on, until it goes through all related cases and their related forms.
+    This recursion is capped at the MAX_CASE_COUNT and MAX_SUBCASE_DEPTH values defined above.
+
+    :param case_obj: The main case object targeted for deletion.
+    :return: Returns 2 dictionaries: one with the complete list of cases and forms to reference for soft deletion
+    once the user confirms the action and another with organized dictionaries of case data to be parsed by the
+    html template. The latter will be further modified in format_case_data_for_display.
+    """
+    @define
+    class DeleteCase:
```

Comment:
could these dataclasses and related utility functions live in a separate utils file outside of this report so that the dataclasses (which are looking great!) and the other related functions don't have to be nested under a larger wrapper function? It's still a bit difficult to digest `get_case_and_display_data` from top to bottom due to all the nesting. It would be great if `walk_case_relations` was further broken down too...

Reply:
perhaps a better place for all these functions could be in a newly created `corehq.apps.hqcase.case_deletion_utils` file?

Reply:
looks like there was also a comment earlier about adding unit tests for these extracted functions...seconding this as well! `hqcase` seems like a more appropriate place for these tests rather that `reports`, so having the utils file there will also make the organization much more digestible for future understanding